src/SessionGenerator.php

```diff
@@ -0,0 +1,303 @@
+     *
+     * @var callable
+     */
+    protected $sessionFactory;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param callable                $sessionFactory Optional session factory callback.
+     * @param int[]|Traversable       $lengths        The session lengths.
+     * @param int                     $padding        Optional padding time.
+     * @param ValidatorInterface|null $validator      Optional validator instance.
+     */
+    public function __construct(
+        callable $sessionFactory,
+        $lengths,
+        $padding = 0,
+        ValidatorInterface $validator = null
+    ) {
+        $this->_setSessionLengths($lengths)
+             ->_setPaddingTime($padding)
+             ->_setValidator($validator)
+             ->_setSessionFactory($sessionFactory);
+    }
+
+    /**
+     * Retrieves the session lengths.
+     *
+     * @since [*next-version*]
+     *
+     * @return int[]|Traversable
+     */
+    protected function _getSessionLengths()
+    {
+        return $this->sessionLengths;
+    }
+
+    /**
+     * Sets the session lengths.
+     *
+     * @since [*next-version*]
+     *
+     * @param int[]|Traversable $sessionLengths
+     *
+     * @return $this
+     */
+    protected function _setSessionLengths($sessionLengths)
+    {
+        $this->sessionLengths = $sessionLengths;
+
+        return $this;
+    }
+
+    /**
+     * Retrieves the padding time.
+     *
+     * @since [*next-version*]
+     *
+     * @return int
+     */
+    protected function _getPaddingTime()
+    {
+        return $this->paddingTime;
+    }
+
+    /**
+     * Sets the padding time.
+     *
+     * @since [*next-version*]
+     *
+     * @param int $paddingTime
+     *
+     * @return $this
+     */
+    protected function _setPaddingTime($paddingTime)
+    {
+        $this->paddingTime = $paddingTime;
+
+        return $this;
+    }
+
+    /**
+     * Retrieves the validator.
+     *
+     * @since [*next-version*]
+     *
+     * @return ValidatorInterface
+     */
+    protected function _getValidator()
+    {
+        return $this->validator;
+    }
+
+    /**
+     * Sets the validator.
+     *
+     * @since [*next-version*]
+     *
+     * @param ValidatorInterface $validator
+     *
+     * @return $this
+     */
+    protected function _setValidator(ValidatorInterface $validator = null)
+    {
+        $this->validator = $validator;
+
+        return $this;
+    }
+
+    /**
+     * Validates a session, if a validator is available.
+     *
+     * @since [*next-version*]
+     *
+     * @param mixed $session
+     *
+     * @return bool
+     */
+    protected function _validateSession($session)
+    {
+        $validator = $this->_getValidator();
+
+        if (is_null($validator)) {
+            return true;
+        }
+
+        try {
+            $validator->validate($session);
+
+            return true;
+        } catch (ValidationFailedExceptionInterface $e) {
```

Comment:
Given that the validation is used to filter which sessions are kept and which are discarded, having the generator throw does not make sense, in my opinion.

The validation _inside_ the generator should allow validation to fail so it can determine that a session is not meant to be yielded. It does this by catching `ValidationFailedExceptionInterface`.

You are right on the name; something like `isSessionValid()` would make more sense.